sklearn/utils/_array_api.py

```diff
@@ -422,47 +505,109 @@ def _add_to_diagonal(array, value, xp):
-        else:
-            scale = sample_score.shape[0]
-        if scale != 0:
-            sample_score = sample_score / scale
+    a = xp.asarray(a, device=device_)
+    if weights is not None:
+        weights = xp.asarray(weights, device=device_)
+
+    output_dtype = None
+    output_dtype_name = None
 
-    if sample_weight is not None:
-        return float(sample_score @ sample_weight)
+    if xp.isdtype(a.dtype, "bool"):
+        a = xp.astype(a, xp.int32)
+    if weights is not None and xp.isdtype(weights.dtype, "bool"):
+        weights = xp.astype(weights, xp.int32)
+
+    if any(
+        (
+            (input_array is not None)
+            and (
+                (not xp.isdtype(input_array.dtype, "numeric"))
+                or xp.isdtype(input_array.dtype, "complex floating")
+            )
+        )
+        for input_array in [a, weights]
+    ):
+        raise ValueError("Expecting only boolean, integral or real floating values.")
+
+    if weights is None and xp.isdtype(a.dtype, "integral"):
+        output_dtype_name = "float64"
+    elif weights is None:
+        output_dtype = a.dtype
+    elif xp.isdtype(a.dtype, "real floating") and xp.isdtype(
+        weights.dtype, "real floating"
+    ):
+        output_dtype = (
+            a.dtype
+            if (xp.finfo(a.dtype).bits >= xp.finfo(a.dtype).bits)
+            else weights.dtype
+        )
     else:
-        return float(xp.sum(sample_score))
+        output_dtype_name = "float64"
+
+    cast_to_float64 = (output_dtype_name == "float64") or (
+        xp.finfo(output_dtype).bits == 64
+    )
+
+    if cast_to_float64 and not _supports_dtype(xp, device_, "float64"):
+        a = _convert_to_numpy(a, xp)
+        if weights is not None:
+            weights = _convert_to_numpy(weights, xp)
+        return xp.asarray(
+            _average(a, axis=axis, normalize=normalize, weights=weights),
+            dtype=xp.float32,
+            device=device_,
+        )
+
+    if output_dtype is None:
+        output_dtype = getattr(xp, output_dtype_name)
+
+    a = xp.astype(a, output_dtype)
+
+    if weights is None:
+        return (xp.mean if normalize else xp.sum)(a, axis=axis)
+
+    weights = xp.astype(weights, output_dtype)
+
+    if a.shape != weights.shape:
+        if axis is None:
+            raise TypeError(
+                "Axis must be specified when shapes of a and weights differ."
+            )
+        if weights.ndim != 1:
+            raise TypeError("1D weights expected when shapes of a and weights differ.")
+
+        if size(weights) != a.shape[axis]:
+            raise ValueError("Length of weights not compatible with specified axis.")
+
+        # If weights are 1D, add singleton dimensions for broadcasting
+        shape = [1] * a.ndim
+        shape[axis] = a.shape[axis]
+        weights = xp.reshape(weights, shape)
+
+    sum_ = xp.sum(xp.multiply(a, weights), axis=axis)
+
+    if not normalize:
+        return sum_
```

Comment:
We could instead divide the weights by `sum_` ahead of the weight multiplications and summing to improve numerical stability.

But we would need to a non-regression test to show that it's actually useful.